### PR TITLE
Espressif ESP32-C3 ESP32-C6 ESP32-S2 Hardware Acceleration

### PR DESCRIPTION
**Pull request by gojimmypi**
# Description

Add Hardware Encryption Acceleration for ESP32-C3, ESP32-C6, and ESP32-S2. 

Feature parity is with original ESP32. There are additional hardware accleration features for some targets to be addressed in a separate PR.

See [Chip Series Comparison](https://docs.espressif.com/projects/esp-idf/en/v5.0/esp32/hw-reference/chip-series-comparison.html)

This PR builds upon the prior work in [#6624](https://github.com/wolfSSL/wolfssl/pull/6624). Included is support for all of the current Espressif ESP-IDF v5.1 targets:

* esp32 - HW acceleration included prior PR
* esp32s2 - HW acceleration included this PR
* esp32c3 - HW acceleration included this PR
* esp32s3 - HW acceleration included prior PR
* esp32c2 - no HW acceleration yet
* esp32c6 - HW acceleration included this PR
* esp32h2 - no HW acceleration yet

See `idf.py --list-targets`. 

With the 5 examples and 7 different targets, there are now 35 builds to check. At some point we may consider moving some (all?) of the examples to [wolfssl-examples](https://github.com/wolfSSL/wolfssl-examples).

In the meantime, I've been testing with a [revised compileAllExamples.sh](https://github.com/gojimmypi/wolfssl/blob/54d951e4b8624dac94bb1ac1492e1b2483225600/IDE/Espressif/ESP-IDF/compileAllExamples.sh#L39) to test all the builds. I need to parallelize it before including it in a PR.

### Not included in this PR:

Although this PR will work with all Espressif targets, not all [wolfSSL examples](https://github.com/wolfSSL/wolfssl/tree/master/IDE/Espressif/ESP-IDF/examples) work without some minor target-specific attention. Rather than muddy up this PR with changes to examples, I'll be submitting a separate PR soon for the minor changes needed. See also pending TLS1.3 examples update in [#6844](https://github.com/wolfSSL/wolfssl/pull/6844).

Fixes zd# n/a/

# Testing

How did you test?

For testing wolfSSL from Linux or WSL:

```bash
git clone https://github.com/gojimmypi/wolfssl.git
cd wolfssl
git checkout PR-Espressif-C3-C6-S2-HW

# Testing Linux
./autogen.sh
./configure --enable-all
git fetch
git pull
git status
make clean && make && ./wolfcrypt/test/testwolfcrypt

# Testing ESP-IDF
cd IDE/Espressif/ESP-IDF
. /mnt/c/SysGCC/esp32/esp-idf/v5.1/export.sh

./compileAllExamples.sh

idf.py flash -p /dev/ttyS35 -b 115200 monitor
```

See [espressif/esp-idf #10423](https://github.com/espressif/esp-idf/issues/10423) for ESP32-C6 development status.

See [#6234](https://github.com/wolfSSL/wolfssl/issues/6234) for Espressif Roadmap and related wolfSSL updates.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h [hunk 1]
```diff
@@ -42,54 +44,76 @@
...
+    ESP_MP_HW_FALLBACK          = (WC_LAST_E - 2),
+    ESP_MP_HW_VALIDATION_ACTIVE = (WC_LAST_E - 3)
+};
 
 /* MP_HW_FALLBACK: signal to caller to fall back to SW for math:
  *   algorithm not supported in SW
  *   known state needing only SW, (e.g. ctx copy)
  *   any other reason to force SW */
-#define MP_HW_FALLBACK (-108)
+#define MP_HW_FALLBACK ESP_MP_HW_FALLBACK  /* was -108 */
```
Comment (bandi13) [inline wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h:57]: Having comments in macro definitions can have unusual unintended consequences. Recommend instead putting it into the comment block above.

**Comment by gojimmypi** [comment]
Jenkins retest this please.

**Comment by gojimmypi** [comment]
Jenkins retest this please.

File: wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h (hunk 1 above)
Comment (gojimmypi) [reply wolfssl/wolfcrypt/port/Espressif/esp32-crypt.h]: Good point on the macro comments. I've updated those and adjusted some line lengths.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/port/Espressif/README.md [hunk 2]
```diff
@@ -8,10 +8,11 @@ For detail about ESP32 HW Acceleration, you can find in [Technical Reference Man
...
 ### Building
 
-To enable hw acceleration :
+```
+git clone --recurse-submodules -j8 https://github.com/espressif/esp-idf master
+```
 
-* Uncomment out `#define WOLFSSL_ESPIDF` in `/path/to/wolfssl/wolfssl/wolfcrypt/settings.h`
-* Uncomment out `#define WOLFSSL_ESP32` in `/path/to/wolfssl/wolfssl/wolfcrypt/settings.h`
+Hardware acceleration is enabled by default.
```
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/src/port/Espressif/README.md]: Ah yes, I missed that README file. I've corrected that and added more detail.

File: wolfcrypt/src/sha.c [hunk 3]
```diff
@@ -630,21 +632,43 @@ int wc_ShaUpdate(wc_Sha* sha, const byte* data, word32 len)
...
 
         if (sha->buffLen == WC_SHA_BLOCK_SIZE) {
-        #if defined(LITTLE_ENDIAN_ORDER) && !defined(FREESCALE_MMCAU_SHA)
-            ByteReverseWords(sha->buffer, sha->buffer, WC_SHA_BLOCK_SIZE);
-        #endif
-
         #if defined(WOLFSSL_USE_ESP32_CRYPT_HASH_HW)
             if (sha->ctx.mode == ESP32_SHA_INIT) {
-                ESP_LOGV(TAG, "wc_ShaUpdate try hardware");
+                #if defined(DEBUG_WOLFSSL_SHA_MUTEX)
```
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/src/sha.c]: Agreed. I could not find a similar "mutex tracking" macro, so I renamed this one to a generic `WOLFSSL_DEBUG_MUTEX`.

File: wolfcrypt/src/sha256.c [hunk 4]
```diff
@@ -701,6 +701,54 @@ static int InitSha256(wc_Sha256* sha256)
...
+     *  see page 337 of C3 spec: 16.4.1.3 Setting the Initial Hash Value
+     *
+     * "Before hash task begins for any secure hash algorithms, the initial
+     * Hash value H(0) must be set based on different algorithms. However,
+     * the SHA accelerator uses the initial Hash values (constant C) stored
+     * in the hardware for hash tasks"
+     */
+    static int set_default_digest256(wc_Sha256* sha256)
+    {
+        return 0;  /* TODO not used? */
```
[the remark this answers is not in the document]
Comment (gojimmypi) [reply wolfcrypt/src/sha256.c]: Oops. My bad, That's code meant for future optimizations. Removed.

**Review by dgarske — DISMISSED** [review]
Looks good to me.

**Comment by gojimmypi** [comment]
Jenkins retest this please.

**Review by JacobBarthelmeh — APPROVED** [review]

**Comment by JacobBarthelmeh** [comment]
Assigning to @gojimmypi for merge conflict and had one question. Looks good otherwise, thanks!

**Comment by gojimmypi** [comment]
Back to you @JacobBarthelmeh :)  merge conflict resolved. Thank you

File: wolfcrypt/src/port/Espressif/esp32_sha.c [hunk 5]
```diff
@@ -138,13 +276,48 @@ int esp_sha_init(WC_ESP32SHA* ctx, enum wc_HashType hash_type)
...
     /* other chipsets will be implemented here */
-#endif /* defined(CONFIG_IDF_TARGET_ESP32) || defined(CONFIG_IDF_TARGET_ESP32S3) */
+    ESP_LOGW(TAG, "SW Fallback; CONFIG_IDF_TARGET = %s", CONFIG_IDF_TARGET);
+    ctx->mode = ESP32_SHA_SW;
+#endif /* CONFIG_IDF_TARGET_ESP32 || x_ESP32S2 || x_ESP32S3 */
 
     return ret;
 }
 
+#ifndef NO_SHAx /* TODO cannot currently turn off SHA */
```
[the remark this answers is not in the document]
Comment (JacobBarthelmeh) [reply wolfcrypt/src/port/Espressif/esp32_sha.c:320]: intentional, NO_SHA support coming later

**Review by JacobBarthelmeh — APPROVED** [review]

**Comment by gojimmypi** [comment]
See https://github.com/wolfSSL/wolfssl/pull/6844 for Espressif example updates to fully utilize this PR.

**Pull request merged**